File: engines/go/internal/background/background.go
package background

import (
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"syscall"

	"statusline/internal/cache"
)

// SpawnCumulativeStats runs cumulative-stats.sh in the background (detached).
func SpawnCumulativeStats(projectDir string) {
	if projectDir == "" {
		return
	}

	exe, err := os.Executable()
	if err != nil {
		return
	}
	selfDir := filepath.Dir(exe)

	// Look for cumulative-stats.sh relative to binary, then fallback paths
	candidates := []string{
		filepath.Join(selfDir, "..", "bash", "cumulative-stats.sh"),
		filepath.Join(selfDir, "cumulative-stats.sh"),
		filepath.Join(os.Getenv("HOME"), ".claude", "cumulative-stats.sh"),
	}

	var script string
	for _, c := range candidates {
		abs, err := filepath.Abs(c)
		if err != nil {
			continue
		}
		if info, err := os.Stat(abs); err == nil && !info.IsDir() {
			script = abs
			break
		}
	}

	if script == "" {
		return
	}

	cmd := exec.Command(script, projectDir)
	cmd.Stdout = nil
	cmd.Stderr = nil
	cmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}
	_ = cmd.Start()
}

// SpawnModelRefresh re-executes the binary with --internal-refresh-models to update the model cache.
func SpawnModelRefresh(sessionID, transcriptPath string) {
	if sessionID == "" || transcriptPath == "" {
		return
	}

	exe, err := os.Executable()
	if err != nil {
		return
	}

	cmd := exec.Command(exe,
		"--internal-refresh-models",
		"--session-id", sessionID,
		"--transcript-path", transcriptPath,
	)
	cmd.Stdin = strings.NewReader("{}")
	cmd.Stdout = nil
	cmd.Stderr = nil
	cmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}
	_ = cmd.Start()
}

// RefreshModelCache is the internal mode that reads JSONL transcripts and writes the model cache.
func RefreshModelCache(sessionID, transcriptPath string) {
	if sessionID == "" || transcriptPath == "" {
		return
	}

	cacheDir := cache.CacheDir()
	_ = os.MkdirAll(cacheDir, 0o755)

	// Collect files to scan
	files := []string{transcriptPath}
	subagentDir := filepath.Join(filepath.Dir(transcriptPath), sessionID, "subagents")
	if entries, err := os.ReadDir(subagentDir); err == nil {
		for _, e := range entries {
			if strings.HasSuffix(e.Name(), ".jsonl") {
				files = append(files, filepath.Join(subagentDir, e.Name()))
			}
		}
	}

	type modelAgg struct {
		Model string `json:"model"`
		In    int    `json:"in"`
		Out   int    `json:"out"`
	}

	models := map[string]*modelAgg{}

	for _, fpath := range files {
		data, err := os.ReadFile(fpath)
		if err != nil {
			continue
		}
		for _, line := range strings.Split(string(data), "\n") {
			line = strings.TrimSpace(line)
			if line == "" {
				continue
			}
			var entry struct {
				Type    string `json:"type"`
				Message struct {
					Model string `json:"model"`
					Usage struct {
						InputTokens             int `json:"input_tokens"`
						OutputTokens            int `json:"output_tokens"`
						CacheReadInputTokens    int `json:"cache_read_input_tokens"`
						CacheCreationInputTokens int `json:"cache_creation_input_tokens"`
					} `json:"usage"`
				} `json:"message"`
			}
			if err := json.Unmarshal([]byte(line), &entry); err != nil {
				continue
			}
			if entry.Type != "assistant" {
				continue
			}
			name := entry.Message.Model
			if name == "" || !strings.HasPrefix(name, "claude-") {
				continue
			}
			if _, ok := models[name]; !ok {
				models[name] = &modelAgg{Model: name}
			}
			models[name].In += entry.Message.Usage.InputTokens +
				entry.Message.Usage.CacheReadInputTokens +
				entry.Message.Usage.CacheCreationInputTokens
			models[name].Out += entry.Message.Usage.OutputTokens
		}
	}

	result := struct {
		Models []*modelAgg `json:"models"`
	}{}
	for _, m := range models {
		result.Models = append(result.Models, m)
	}

	data, err := json.Marshal(result)
	if err != nil {
		return
	}

	cacheFile := filepath.Join(cacheDir, fmt.Sprintf("models-%s.json", sessionID))
	tmp := cacheFile + ".tmp"
	if err := os.WriteFile(tmp, data, 0o644); err != nil {
		return
	}
	_ = os.Rename(tmp, cacheFile)
}

File: engines/rust/src/background.rs
use crate::cache;
use std::env;
use std::fs;
use std::os::unix::process::CommandExt;
use std::path::{Path, PathBuf};
use std::process::{Command, Stdio};

/// Run cumulative-stats.sh in the background (detached).
pub fn spawn_cumulative_stats(project_dir: &str) {
    if project_dir.is_empty() {
        return;
    }

    let exe = match env::current_exe() {
        Ok(e) => e,
        Err(_) => return,
    };
    let self_dir = match exe.parent() {
        Some(d) => d.to_path_buf(),
        None => return,
    };

    let candidates = [
        self_dir.join("..").join("bash").join("cumulative-stats.sh"),
        self_dir.join("cumulative-stats.sh"),
        PathBuf::from(env::var("HOME").unwrap_or_default())
            .join(".claude")
            .join("cumulative-stats.sh"),
    ];

    let mut script = None;
    for c in &candidates {
        if let Ok(abs) = fs::canonicalize(c) {
            if abs.is_file() {
                script = Some(abs);
                break;
            }
        }
    }

    let script = match script {
        Some(s) => s,
        None => return,
    };

    let mut cmd = Command::new(&script);
    cmd.arg(project_dir);
    cmd.stdout(Stdio::null());
    cmd.stderr(Stdio::null());
    cmd.stdin(Stdio::null());
    unsafe {
        cmd.pre_exec(|| {
            libc::setpgid(0, 0);
            Ok(())
        });
    }
    let _ = cmd.spawn();
}

/// Re-execute the binary with --internal-refresh-models to update the model cache.
pub fn spawn_model_refresh(session_id: &str, transcript_path: &str) {
    if session_id.is_empty() || transcript_path.is_empty() {
        return;
    }

    let exe = match env::current_exe() {
        Ok(e) => e,
        Err(_) => return,
    };

    let mut cmd = Command::new(&exe);
    cmd.args([
        "--internal-refresh-models",
        "--session-id",
        session_id,
        "--transcript-path",
        transcript_path,
    ]);
    cmd.stdin(Stdio::null());
    cmd.stdout(Stdio::null());
    cmd.stderr(Stdio::null());
    unsafe {
        cmd.pre_exec(|| {
            libc::setpgid(0, 0);
            Ok(())
        });
    }
    let _ = cmd.spawn();
}

/// Parse JSONL transcripts and write the model cache (internal mode).
pub fn refresh_model_cache(session_id: &str, transcript_path: &str) {
    if session_id.is_empty() || transcript_path.is_empty() {
        return;
    }

    let cache_dir = cache::cache_dir();
    let _ = fs::create_dir_all(&cache_dir);

    // Collect files to scan
    let mut files = vec![PathBuf::from(transcript_path)];
    let subagent_dir = Path::new(transcript_path)
        .parent()
        .map(|p| p.join(session_id).join("subagents"))
        .unwrap_or_default();

    if let Ok(entries) = fs::read_dir(&subagent_dir) {
        for entry in entries.flatten() {
            let name = entry.file_name();
            if name.to_string_lossy().ends_with(".jsonl") {
                files.push(entry.path());
            }
        }
    }

    #[derive(serde::Serialize)]
    struct ModelAgg {
        model: String,
        #[serde(rename = "in")]
        in_tokens: i64,
        out: i64,
    }

    let mut models: std::collections::HashMap<String, ModelAgg> = std::collections::HashMap::new();

    for fpath in &files {
        let data = match fs::read_to_string(fpath) {
            Ok(d) => d,
            Err(_) => continue,
        };
        for line in data.lines() {
            let line = line.trim();
            if line.is_empty() {
                continue;
            }
            let entry: serde_json::Value = match serde_json::from_str(line) {
                Ok(v) => v,
                Err(_) => continue,
            };
            if entry.get("type").and_then(|t| t.as_str()) != Some("assistant") {
                continue;
            }
            let msg = match entry.get("message") {
                Some(m) => m,
                None => continue,
            };
            let name = match msg.get("model").and_then(|m| m.as_str()) {
                Some(n) if n.starts_with("claude-") => n.to_string(),
                _ => continue,
            };
            let usage = match msg.get("usage") {
                Some(u) => u,
                None => continue,
            };
            let input = usage
                .get("input_tokens")
                .and_then(|v| v.as_i64())
                .unwrap_or(0)
                + usage
                    .get("cache_read_input_tokens")
                    .and_then(|v| v.as_i64())
                    .unwrap_or(0)
                + usage
                    .get("cache_creation_input_tokens")
                    .and_then(|v| v.as_i64())
                    .unwrap_or(0);
            let output = usage
                .get("output_tokens")
                .and_then(|v| v.as_i64())
                .unwrap_or(0);

            let agg = models.entry(name.clone()).or_insert(ModelAgg {
                model: name,
                in_tokens: 0,
                out: 0,
            });
            agg.in_tokens += input;
            agg.out += output;
        }
    }

    #[derive(serde::Serialize)]
    struct Result {
        models: Vec<ModelAgg>,
    }

    let result = Result {
        models: models.into_values().collect(),
    };

    let data = match serde_json::to_vec(&result) {
        Ok(d) => d,
        Err(_) => return,
    };

    let cache_file = cache_dir.join(format!("models-{}.json", session_id));
    let tmp = cache_file.with_extension("json.tmp");
    if fs::write(&tmp, &data).is_ok() {
        let _ = fs::rename(&tmp, &cache_file);
    }
}
